Chore: Upgrade typescript to 3.7

kaydelaney opened the pull request:
**What this PR does / why we need it**:
Upgrades typescript to the latest version.

kaydelaney committed:
Chore: Upgrade typescript to 3.7

dprokop commented on packages/grafana-ui/src/slate-plugins/suggestions.tsx:
```diff
@@ -6,14 +6,7 @@ import { Editor as CoreEditor } from 'slate';
-  SuggestionsState,
-  CompletionItemGroup,
-} from '..';
+import { makeFragment, TypeaheadOutput, CompletionItem, TypeaheadInput, SuggestionsState } from '..';
```
Could you please update this import to be from the accurate path rather than `..`? We also need to introduce linter for such imports as they are problematic for Storybook.

kaydelaney replied:
Sure

dprokop commented on public/app/features/plugins/datasource_srv.ts:
```diff
@@ -17,7 +17,7 @@ import { GrafanaRootScope } from 'app/routes/GrafanaCtrl';
 
 export class DatasourceSrv implements DataSourceService {
-  datasources: { [name: string]: DataSourceApi };
+  datasources: Record<string, DataSourceApi>;
```
Hm, never seen `Record` before. Is this a new thing introduced in TS3.7? We also have wildly used  `KeyValue` type in grafana/data which could probably be replaced with `Record`

kaydelaney replied:
It's existed for a while as far as I know (https://www.typescriptlang.org/docs/handbook/utility-types.html#recordkt). Mostly a personal preference thing but I think it reads a bit better than `{ [key: string]: DataSourceApi }`

dprokop commented on packages/grafana-data/src/types/datasource.ts:
```diff
@@ -5,18 +5,32 @@ import { PanelData } from './panel';
   onOptionsChange: (options: DataSourceSettings<JSONData, SecureJSONData>) => void;
 }
 
+export type DataSourceQuery<DSType extends DataSourceApi<any, any>> = DSType extends DataSourceApi<
```
@kaydelaney can you say something more about these changes? Can't see how they fit into TS upgrade.

cc @ryantxu

kaydelaney replied:
Sure. Typescript 3.7 is somewhat more strict with its typechecking in certain scenarios, which caused a type error when doing the following
```typescript
const plugin = new DataSourcePlugin(lokiDatasource);
```
with the error
```
"Argument of type 'typeof LokiDatasource' is not assignable to parameter of type 'DataSourceConstructor<DataSourceApi<DataQuery, LokiOptions>, DataQuery, LokiOptions>'."
```
So in order for the correct types to be properly inferred the helper types above were introduced to extract the `TQuery` and `TOptions` types of a type extending `DataSourceApi<TQuery, TOptions>`.

dprokop commented on packages/grafana-data/src/types/datasource.ts:
```diff
@@ -284,6 +298,7 @@ export interface ExploreQueryFieldProps<
 > extends QueryEditorProps<DSType, TQuery, TOptions> {
   history: any[];
   onBlur?: () => void;
+  absoluteRange?: AbsoluteTimeRange;
```
Why is this added here? 

cc @ryantxu

kaydelaney replied:
An `absoluteRange` prop is passed to `QueryField` components at some points (see QueryRow.tsx, line 168), which causes an error in typescript 3.7 since `absoluteRange` wasn't a member of the props interface.

Full error:
```
Type '{ datasource: DataSourceApi<DataQuery, DataSourceJsonData>; query: DataQuery; history: HistoryItem<DataQuery>[]; ... 5 more ...; absoluteRange: AbsoluteTimeRange; }' is not assignable to type 'IntrinsicAttributes & ExploreQueryFieldProps<DataSourceApi<DataQuery, DataSourceJsonData>, DataQuery, DataSourceJsonData> & { ...; }'.
  Property 'absoluteRange' does not exist on type 'IntrinsicAttributes & ExploreQueryFieldProps<DataSourceApi<DataQuery, DataSourceJsonData>, DataQuery, DataSourceJsonData> & { ...; }'.ts(2322)
```

dprokop commented on packages/grafana-ui/src/components/Chart/Tooltip.test.tsx:
```diff
@@ -12,7 +12,7 @@ describe('Chart Tooltip', () => {
       // when measuring tooltip container (wrapper with padding and content inside)
       Element.prototype.getBoundingClientRect = jest.fn(() => {
-        return { width: 100, height: 100, top: 0, left: 0, bottom: 0, right: 0 };
+        return { width: 100, height: 100, top: 0, left: 0, bottom: 0, right: 0 } as any;
```
`as ClientRect | DOMRect` ?

kaydelaney replied:
good idea

dprokop commented:
Added @ryantxu as I see some changes he might be interested in

kaydelaney committed:
Small improvements based on feedback
- Changes import paths in suggestions.tsx
- Changes type of DOMRect in Tooltip.test.tsx from any to DOMRect

ryantxu commented on packages/grafana-data/src/types/datasource.ts:
```diff
@@ -5,18 +5,32 @@ import { PanelData } from './panel';
+>
+  ? TOptions
+  : never;
+
```
**head... exploding**

I'll trust you on this one :)  The nested typing on the datasources getting super hard to follow - I think that is just the nature of the structure, but is something we should try to clean up if possible

kaydelaney replied:
Agreed

ryantxu reviewed (approved):
Looks good -- the DataSourceAPI typing is complex, but that is not new :(

kaydelaney committed:
Adds some comments and simplifies some types

dprokop reviewed (approved):
@kaydelaney thanks for explaining all the things, great stuff!
